discovery_engine_core/core/src/engine.rs

```diff
@@ -810,18 +798,29 @@ async fn update_stacks<'a>(
         return Ok(());
     }
 
-    let key_phrases = needy_stacks
+    let key_phrases_by_market = needy_stacks
```

Comment:
❓ it seems a bit unfortunate that we have to collect a hashmap in this first loop just to pick the entries one by one in the second loop below. i get that we can't just move the `ranker.take_key_phrases()` into the second loop because the captured `&mut ranker` can't be passed along with the constructed future. instead of the current `join_all(Vec<Future>)`, would this maybe work with sth like a `FuturesUnordered<_>` collection? or if we set up the collecting and joining of the futures differently eg. with a join macro?